pki/denylist.go

```diff
@@ -0,0 +1,253 @@
+	if b.URL() == "" {
+		return nil
+	}
+
+	// Extract the necessary information from the certificate in order to process the denylist entry
+	issuer := cert.Issuer.String()
+	serialNumber := cert.SerialNumber.String()
+	thumbprint := certKeyJWKThumbprint(cert)
+
+	// If the denylist has not yet been downloaded, do so now
+	if b.lastUpdated.IsZero() {
+		// Trigger an update of the denylist
+		if err := b.Update(); err != nil {
+			// If the denylist download failed then log a message about it
+			logger().WithError(err).
+				WithField("Issuer", issuer).
+				WithField("S/N", serialNumber).
+				Error("Cert validation failed because the denylist cannot be downloaded")
+
+			// Return an error indicating the denylist cannot be retrieved
+			return ErrDenylistMissing
+		}
+	}
+
+	// Fetch the entries
+	entriesPtr := b.entries.Load()
+
+	// The entries pointer should not be empty because of the lastUpdated check above
+	if entriesPtr == nil {
+		// If the entries still cannot be fetched then something is not right so return an error
+		logger().Error("BUG: denylist entries pointer is nil")
+		return ErrDenylistMissing
+	}
+
+	// Check each entry in the denylist for matches
+	for _, entry := range *entriesPtr {
+		// Check for this issuer and serial number combination
+		if entry.Issuer == issuer && entry.SerialNumber == serialNumber && entry.JWKThumbprint == thumbprint {
+			// Return an error indicating the certificate has been denylisted
+			return fmt.Errorf("%w: %s", ErrCertBanned, entry.Reason)
+		}
+	}
+
+	// Return a nil error as the certificate hasn't been denylisted
+	return nil
+}
+
+func (b *denylistImpl) LastUpdated() time.Time {
+	return b.lastUpdated
+}
+
+func (b *denylistImpl) URL() string {
+	return b.url
+}
+
+// update downloads the denylist, and updates the in-memory representation
+func (b *denylistImpl) Update() error {
+	// Updating a denylist with a URL is a NOP
+	if b.URL() == "" {
+		return nil
+	}
+
+	// Download the denylist
+	bytes, err := b.download()
+	if err != nil {
+		logger().WithError(err).
+			WithField("URL", b.url).
+			Warn("Certiciate denylist cannot be downloaded")
+		return err
+	}
+
+	// Check the signature of the denylist
+	payload, err := jws.Verify(bytes, jwa.EdDSA, b.trustedKey)
+	if err != nil {
+		return fmt.Errorf("failed to verify denylist signature: %w", err)
+	}
+
+	// Parse the JSON of the payload
+	var entries []denylistEntry
+	if err = json.Unmarshal(payload, &entries); err != nil {
+		return fmt.Errorf("failed to parse denylist JSON: %w", err)
+	}
+
+	// Store the new denylist entries
+	b.entries.Store(&entries)
+
+	// Track when the denylist was last updated
+	b.lastUpdated = time.Now()
+
+	// Return a nil error as the denylist was successfully updated
+	return nil
+}
+
+// download retrieves and parses the denylist
+func (b *denylistImpl) download() ([]byte, error) {
+	// Make an HTTP GET request for the denylist URL
+	response, err := http.Get(b.url)
```

Comment:
Needs a timeout (e.g. use a context), otherwise `download()` could block (which uses a waitgroup)?

Reply:
now uses the same timeout as the crl validator (10 sec)